### PR TITLE
Fix Issue 18775, 12807: gently resolve alias this

### PR DESCRIPTION
**Pull request by FeepingCreature**
Only emit warnings during alias this resolution if the alias this later actually turns out to lead to a valid property access.

This fixes issues where the impropriety of the alias this is reported as warnings or errors, even if the alias this doesn't actually end up being taken.

The underlying problem is that alias this competes with UFCS for property resolution, and is evaluated first. So UFCS to an alias this type always triggers the alias this, even if it doesn't match call types.

~~Maybe the better course of action would be to check UFCS before alias this, but I don't know how to do that.~~ No, alias this should correctly take priority.

### CODE REVIEW

**Comment by FeepingCreature** [comment]
How do I get the bot to track two issues...?

File: test/compilable/test12807.d
Comment (RazvanN7) [inline test/compilable/test12807.d]: please add the test case from Issue 18775 also.

**Review by RazvanN7 — COMMENTED** [review]
I am not convinced that the alias this should not report the deprecation. FWIW aliasing this a deprecated method implies that the line of code which implements the alias uses the deprecated function. I would argue that this calls for a deprecation message even if the alias this does not actually get to be used and it should be reported at the line where the alias this is created to a deprecated symbol.

File: test/compilable/test12807.d
Comment (FeepingCreature) [reply test/compilable/test12807.d]: How do I add a testcase that ensures that a deprecation message does not get printed?

**Comment by FeepingCreature** [comment]
Unfortunately, at present deprecating the alias itself does nothing. However, even so, precedent seems to be that a deprecated function only triggers a warning when it is actually used.

In my opinion, an alias of a function should not in itself qualify as a use.

File: test/compilable/test12807.d
Comment (RazvanN7) [reply test/compilable/test12807.d]: add a comment like this : 

// REQUIRED_ARGS: -de

It tells the tester to treat deprecations as errors

Comment (FeepingCreature) [reply test/compilable/test12807.d]: Added.

**Comment by jacob-carlborg** [comment]
The commits say: "Mathis Beer authored and Mathis Beer committed". It would be great if you could commit as the same user as the GitHub account.

**Comment by FeepingCreature** [comment]
I'm working on this from my work pc, which is configured for my irl name. I'm not actually sure how to change my git identity; however, if anything, my irl name is more appropriate because it's work I'm doing "on company time."

**Comment by jacob-carlborg** [comment]
I think it's based on the email address and the name used to commit. See https://help.github.com/articles/setting-your-commit-email-address-in-git
https://help.github.com/articles/setting-your-username-in-git/

**Comment by jacob-carlborg** [comment]
You can also add your real name to your GitHub account without changing the account name. Have a look at my profile: https://github.com/jacob-carlborg

**Comment by jacob-carlborg** [comment]
Even though you're not committing using the same email address as your GitHub account, there should be no reason for the committer and author to be different.

**Comment by FeepingCreature** [comment]
Aah! The problem is that I really want to keep this _account_ largely separate from my work, but I also don't want to make a new Github account just for work. It's a bit awkward.

There should be no reason for the committer and author to be different. In fact, they are not different. I looked at the commit; they're exactly the same. I have no idea why Github shows them separately.

**Comment by jacob-carlborg** [comment]
> There should be no reason for the committer and author to be different. In fact, they are not different. I looked at the commit; they're exactly the same. I have no idea why Github shows them separately.

Hmm, yes. Strange that GitHub thinks the committer and author are not the same.

**Comment by FeepingCreature** [comment]
I think that semaphoreci failure is spurious.

**Comment by FeepingCreature** [comment]
Yep!

**Comment by FeepingCreature** [comment]
How do I request review again?

File: src/dmd/typesem.d
Comment (RazvanN7) [inline src/dmd/typesem.d]: Why is this necessary? Won't the errors be reported at line 3104?

Comment (RazvanN7) [reply src/dmd/typesem.d]: It it's necessary, why is `if (!gagError)` necessary?

Comment (FeepingCreature) [reply src/dmd/typesem.d]: No, because we're bailing here - we can't continue because alias_e is null, so we can't make a DotIdExp of it. This can happen if the target of the alias is itself invalid. This is the part that fixes Issue 12807.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: I guess I could set e to the ungagged expression because it'll know that it's erroring.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: No I can't! Because even so, the later property access could still fail a la 18775. So I cannot error that early ever, anyways.

I stand by the current impl.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: ... No, we can keep going if we're not gagged. Because we'll be gagged via flag for the test to decide whether we can do ufcs after? ....? ???

Comment (FeepingCreature) [reply src/dmd/typesem.d]: Oh, fuck. Strictly speaking I can only emit deprecations if they are *not* errors.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: Because if they *are* errors, we may want to fall through into ufcs?

Comment (FeepingCreature) [reply src/dmd/typesem.d]: ... I'll stick my neck out and say that the code as written _possibly_ does the Right Thing.

I think it definitely does a significantly righter thing than DMD does at present.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: Hm. Maybe I've been coming at this wrong.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: Okay! Here's the __issue__.

The problem is **warnings**. If I get an error in the alias resolve, I can just keep going with the errored exp; if it returns null, I can just return null since we'd be gagged. The problem is if the initial alias resolve prints a **warning**, then later on the dot lookup **fails** and we fall through to UFCS, we'll have printed a warning spuriously. That's the case that I have to workaround with the repeated resolveAliasThis calls.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: I actually managed to find an improvement.

This code is fucking *hellish*.

**Review by RazvanN7 — APPROVED** [review]
Thanks for the effort. Looks good to me. I'll wait a few days before tagging auto-merge to offer the possibility to other reviewers to jump in.

**Comment by FeepingCreature** [comment]
Cheers. <3

**Comment by RazvanN7** [comment]
Looks like you shot 2 "rabbits" on your first shot in dmd :drum:

**Comment by FeepingCreature** [comment]
Third or fourth shot, actually! And yeah, somebody pointed it out in the comments of the bug report, thankfully.

File: src/dmd/typesem.d
Comment (ibuclaw) [inline src/dmd/typesem.d]: Makes sense, but pretty awful that you have to call this twice.

Comment (FeepingCreature) [reply src/dmd/typesem.d]: I agree. I don't know a better way though.

**Comment by FeepingCreature** [comment]
Monday morning ping! :-)

**Pull request merged**
